In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import glob
import shutil
import os

In [2]:
metadata_df = pd.read_csv("C:/Users/abhij/Downloads/metadata.csv")

In [3]:
metadata_df.head(5)


Record ID                                          File Name Assignment ID  \
0  207431327         HearstTransportation_CAMP_MontrealTeam.png      CPR6931p   
1  207431332  HearstTransportation_MicrosoftTeams_ZoomMeetin...      CPR6931p   
2  207431348   HearstTransportation_BlackBook_LauraWehunt_1.png      CPR6931p   
3  207431356   HearstTransportation_BlackBook_LauraWehunt_2.png      CPR6931p   
4  207432753                   HearstMagazines_CDS_Employee.png      CPR6931p   

  Year                                             Person  \
0  NaN                                       Unidentified   
1  NaN  Unidentified,Don Dickson,Danya Bynoe,Varun Kum...   
2  NaN                                       Laura Wehunt   
3  NaN                                       Laura Wehunt   
4  NaN                                     Bryan Phillips   

                                            Keywords Creators Date  
0  People @ Work,Hearst transportation,Transporta...      NaN  NaN  
1  People @ Work,Hearst transportation,Transporta...      NaN  NaN  
2  People @ Work,Hearst transportation,Headshots,...      NaN  NaN  
3  People @ Work,Hearst transportation,Headshots,...      NaN  NaN  
4  People @ Work,CDS,Hearst Magazines,Hearst maga...      NaN  NaN

In [4]:
metadata_df.groupby(by = ['Keywords']).agg('count').sort_values(by = ['Record ID'], ascending=False).head(30)

Record ID  File Name  \
Keywords                                                                   
People @ Work,Homecoming,2013,Events,44th Floor...        443        443   
People @ Work,2018,Events,Friday,Texas,Texas 20...        258        258   
Archdiocese of New York,Corporate,Al Smith Dinn...        230        230   
People @ Work,Spotlight 2016,Awards,Hearst Maga...        140        140   
People @ Work,Frank A. Bennack,Frank Bennack,Ev...        127        127   
People @ Work,2018,Thursday,Events,Texas,Texas ...        114        114   
People @ Work,Tower Gala,Events,Hearst Tower              111        111   
People @ Work,2019,Events,Friday,Texas,Texas 2019          98         98   
People @ Work,Texas 2016,2016,Friday,Events,Texas          97         97   
People @ Work,Texas 2017,2017,Events,Friday,Texas          96         96   
Scenery,Hearst Real Estate,2021,Outdoors,San Si...         94         94   
People @ Work,Culture Shifting Weekend,Culture ...         90         90   
People @ Work,2019,Leave Something on the Table...         89         89   
44th Floor,Events,Hearst Ventures,Ventures,Ken ...         86         86   
People @ Work,2019,Thursday,Events,Texas,Texas ...         76         76   
Hearst Real Estate,2021,Outdoors,San Simeon,Hor...         73         73   
People @ Work,2019,Awards,Hearst Magazines,Even...         72         72   
Cooking,Jack Ranch,Hearst Real Estate,2021,Hear...         68         68   
People @ Work,Texas 2016,2016,Thursday,Events,T...         66         66   
People @ Work,Texas 2016,Alison Krauss,2016,Fri...         65         65   
People @ Work,2018,Awards,Hearst Magazines,Even...         64         64   
People @ Work,2017,Awards,Hearst Magazines,Even...         62         62   
People @ Work,2018,Frank A. Bennack,Thursday,Fr...         61         61   
People @ Work,Editorial Excellence,Hearst Magaz...         60         60   
Frank Bennack,FAB,Homecoming,44th Floor,People ...         60         60   
People @ Work,Owl's Nest,Event,44th Floor                  59         59   
People @ Work,Texas 2017,2017,Thursday,Events,T...         58         58   
People @ Work,Editorial Excellence,2018,Hearst ...         56         56   
People @ Work,Texas 2017,2017,Frank A. Bennack,...         55         55   
Hearst.com,Headshots,1276x960,Portraits,HearstL...         54         54   

                                                    Assignment ID  Year  \
Keywords                                                                  
People @ Work,Homecoming,2013,Events,44th Floor...            443   443   
People @ Work,2018,Events,Friday,Texas,Texas 20...            258   258   
Archdiocese of New York,Corporate,Al Smith Dinn...              0   230   
People @ Work,Spotlight 2016,Awards,Hearst Maga...              0   140   
People @ Work,Frank A. Bennack,Frank Bennack,Ev...            127   127   
People @ Work,2018,Thursday,Events,Texas,Texas ...            114   114   
People @ Work,Tower Gala,Events,Hearst Tower                  111   111   
People @ Work,2019,Events,Friday,Texas,Texas 2019              98    98   
People @ Work,Texas 2016,2016,Friday,Events,Texas              97    97   
People @ Work,Texas 2017,2017,Events,Friday,Texas              96    96   
Scenery,Hearst Real Estate,2021,Outdoors,San Si...             94    94   
People @ Work,Culture Shifting Weekend,Culture ...             89    90   
People @ Work,2019,Leave Something on the Table...              0    89   
44th Floor,Events,Hearst Ventures,Ventures,Ken ...              0    86   
People @ Work,2019,Thursday,Events,Texas,Texas ...             76    76   
Hearst Real Estate,2021,Outdoors,San Simeon,Hor...             73    73   
People @ Work,2019,Awards,Hearst Magazines,Even...              0    72   
Cooking,Jack Ranch,Hearst Real Estate,2021,Hear...             68    68   
People @ Work,Texas 2016,2016,Thursday,Events,T...             66    66   
People @ Work,Texas 2016,Alison Krauss,2016,

## Feature Engineering

#### Split on comma and find individual keywords

In [5]:
clean_df = metadata_df.copy()

In [6]:
clean_df = clean_df.join(clean_df['Keywords'].str.get_dummies(','))

In [7]:
tag_df = pd.DataFrame(clean_df.iloc[:,7:].sum(axis = 0), columns = ['Count']).sort_values(by = ['Count'], ascending = False)

In [8]:
tag_df['Tag_Name'] = tag_df.index
tag_df.reset_index(drop = True, inplace = True)

In [9]:
### Further steps
tag_df['Count_tag_appears'] = tag_df['Count']
tag_df = tag_df.iloc[:,1:]

In [10]:
# Creating five broad categories
time_list = ['Friday','Thursday','2019','2018','2016','2021','2017','2013']
tag_df.head

<bound method NDFrame.head of               Tag_Name  Count_tag_appears
0        People @ Work               7058
1               Events               6483
2           44th Floor               2993
3                Texas               2090
4     Hearst Magazines               1611
...                ...                ...
1446   Joanne Alfarone                  1
1447      John Francis                  1
1448           Johnson                  1
1449   Jonathan Becker                  1
1450         Portraits                  1

[1451 rows x 2 columns]>

In [11]:
tag_df[tag_df['Count_tag_appears']>100].count()

Tag_Name             116
Count_tag_appears    116
dtype: int64

#### Finding A Subset of Tags
* There are 116 tags that have a count of atleast 100 in the dataset
*Find all tags that have a count of atleast 100. Set this as an assorted list of tags
*Check in clean_df the count of images where any one tag appears which is in the assorted list of tags
*Remove year of picture From this tag list.
*Ultimately reduce tags created as dummy features in original dataset to reduce dimensionality

In [12]:
tag_list = list(tag_df[tag_df['Count_tag_appears']>140]['Tag_Name'])

In [13]:
tag_list

['People @ Work',
 'Events',
 '44th Floor',
 'Texas',
 'Hearst Magazines',
 'Magazines',
 'FAB',
 'Frank Bennack',
 '2019',
 'Frank A. Bennack',
 '2021',
 '2016',
 '2018',
 'Hearst Tower',
 'Friday',
 'Thursday',
 '2013',
 'The Long Gray Line',
 'Homecoming',
 'Retreat',
 'Texas 2018',
 'Corporate',
 'Headshots',
 'Portraits',
 '2017',
 'SRS',
 'Spotlight',
 'Steve Swartz',
 'Steven R. Swartz',
 'Awards',
 'Annual',
 'Hearst magazines',
 'Executives',
 'Alfred E. Smith Memorial Foundation',
 'Al Smith Dinner',
 'Archdiocese of New York',
 'Charities',
 'Alfred E. Smith Memorial Foundation Dinner',
 'Al Smith Dinner 2021',
 'Catholic charities',
 'Dinners',
 'Texas 2016',
 'Editorial Excellence',
 'magazines',
 'Texas 2017',
 'Gil Maurer',
 'GCM',
 'Texas 2019',
 'Masterclass',
 'Gilbert C. Maurer',
 'Annual Review',
 'Leave Something on the Table',
 'Spotlight 2016',
 'Tower Gala',
 'Photography',
 'Hearst Real Estate',
 'Real Estate',
 'Tower',
 'Tower Week',
 'Tower Tuesdays',
 'Jose

In [14]:
#Count of images that can be excluded cannot be more than 0.5% of the dataset size. 0.5% of 9300 images roughly is 46.5. On Trial and Error a cutoff of 140 tags is found.
clean_df[(clean_df[tag_list].any(axis = 1))==False]

Record ID                                          File Name  \
54    207993765                    CPR7566_HTV_2020_YearEnder1.png   
55    207993767                    CPR7566_HTV_2020_YearEnder6.png   
56    207993773                    CPR7566_HTV_2020_YearEnder5.png   
57    207993776       CPR7566_HoustonChron_TogetherWeAreStrong.png   
58    207993778                    CPR7566_HTV_2020_YearEnder7.png   
59    207993781      CPR7566_WPBF_ProjectCommUNITY_DayOfGiving.png   
60    207993784                    CPR7566_HTV_2020_YearEnder2.png   
61    207993797                    CPR7566_HTV_2020_YearEnder3.png   
62    207993798          CPR7566_WWVTM_HurricaneLaura_Met@Desk.png   
63    207993803                    CPR7566_HTV_2020_YearEnder4.png   
66    208015824  CPR7560i_BHM_Article Image_V4_1276x960_Tommy_2...   
1254  207993766    CPR7566_WDSU_HurricaneLaura_TrackingKweilyn.jpg   
1255  207993768         CPR7566_WCVB_CelebrityMashupFundraiser.jpg   
1256  207993769                  CPR7566_WYFF_HurricaneIsaias2.jpg   
1257  207993770   CPR7566_KSBW_ProjectCommUNITY_JarrettSuccess.jpg   
1258  207993771      CPR7566_KCCI_Derecho_SpecialRelief_Stacey.jpg   
1259  207993772           CPR7566_WVTM_HurricaneLaura_Reporter.JPG   
1260  207993774                  CPR7566_KCRA_Wildfires_Brian.jpeg   
1261  207993777                   CPR7566_KSBW_Wildfires_Alani.jpg   
1262  207993779  CPR7566_KCRA_ProjectCommUNITY_Foodbank_Anchors...   
1263  207993780           CPR7566_WDSU_HurricaneLaura_Telethon.jpg   
1264  207993783  CPR7566_WTAE_ProjectCommUNITY_DayOfGiving_Chec...   
1265  207993785            CPR7566_KCCI_Derecho_TotalDonations.jpg   
1266  207993786  CPR7566_WTAE_ProjectCommUNITY_DayOfGiving_Mult...   
1267  207993788             CPR7566_KCCI_Derecho_DamagedHouse.jpeg   
1268  207993789                  CPR7566_KCRA_Wildfires_Vicki.jpeg   
1269  207993792             CPR7566_KCRA_Wildfires_VickiSmoke.jpeg   
1270  207993796   CPR7566_WDSU_HurricaneLaura_TrackingMargaret.jpg   
1271  207993799                  CPR7566_WYFF_HurricaneIsaias1.jpg   
1272  207993801                 CPR7566_KSBW_Wildfires_Caitlin.jpg   
1273  207993802         CPR7566_KCRA_Wildfires_FundraiseTally.jpeg   
1490  208032792  CPR7603_Matter of Fact Listening Tour March_RI...   
1491  208032793  CPR7603_Matter of Fact Listening Tour March_RI...   
1561  208702302                         ana_occhiuzzi_1276x960.jpg   
1562  208702347                              CPR6479B_Katie Hu.jpg   
8133  207386216                        0840_JM2_6419_USSYP2019.jpg   
8134  207386217                        0842_JM2_6423_USSYP2019.jpg   
8135  207386218                        0815_JM1_9779_USSYP2019.jpg   
8138  207386221                        0808_EL2_6745_USSYP2019.jpg   
8157  207386243      368_National_Gallery_Gala_USSYP2012_Day06.jpg   
8171  207386267  1_6351624608504597893245215_25_HCPF_092813_pm_...   
8173  207386269  0_63516204568277625016445215_28_HCPF_092813_pm...   
8177  207386274                                   AnissaBalson.jpg   
8180  207386892                         HEAR1_20120918_PMC_194.jpg   
8181  207386893  Austin+Hearst+2nd+Annual+Save+Children+Illumin...   

     Assignment ID        Year  \
54         CPR7566         NaN   
55         CPR7566         NaN   
56         CPR7566         NaN   
57         CPR7566         NaN   
58         CPR7566         NaN   
59         CPR7566         NaN   
60         CPR7566         NaN   
61         CPR7566         NaN   
62         CPR7566         NaN   
63         CPR7566         NaN   
66         CPR7560  2021-02-26   
1254       CPR7566         NaN   
1255       CPR7566         NaN   
1256       CPR7566         NaN   
1257       CPR7566         NaN   
1258       CPR7566         NaN   
1259       CPR7566         NaN   
1260       CPR7566         NaN   
1261       CPR7566         NaN   
1262       CPR7566         NaN   
1263       CPR7566         NaN   
1264       CPR7566         NaN   
126

In [15]:
#Number of features to be used in dataset for training before feature importance is taken
len(tag_list)+ metadata_df.shape[1] - 1

103

In [16]:
image_df_original = clean_df.iloc[:,[0,1,2,3,4,6,7]]

In [17]:
image_df_original = image_df_original.join(clean_df[tag_list])

In [18]:
(tag_list)

['People @ Work',
 'Events',
 '44th Floor',
 'Texas',
 'Hearst Magazines',
 'Magazines',
 'FAB',
 'Frank Bennack',
 '2019',
 'Frank A. Bennack',
 '2021',
 '2016',
 '2018',
 'Hearst Tower',
 'Friday',
 'Thursday',
 '2013',
 'The Long Gray Line',
 'Homecoming',
 'Retreat',
 'Texas 2018',
 'Corporate',
 'Headshots',
 'Portraits',
 '2017',
 'SRS',
 'Spotlight',
 'Steve Swartz',
 'Steven R. Swartz',
 'Awards',
 'Annual',
 'Hearst magazines',
 'Executives',
 'Alfred E. Smith Memorial Foundation',
 'Al Smith Dinner',
 'Archdiocese of New York',
 'Charities',
 'Alfred E. Smith Memorial Foundation Dinner',
 'Al Smith Dinner 2021',
 'Catholic charities',
 'Dinners',
 'Texas 2016',
 'Editorial Excellence',
 'magazines',
 'Texas 2017',
 'Gil Maurer',
 'GCM',
 'Texas 2019',
 'Masterclass',
 'Gilbert C. Maurer',
 'Annual Review',
 'Leave Something on the Table',
 'Spotlight 2016',
 'Tower Gala',
 'Photography',
 'Hearst Real Estate',
 'Real Estate',
 'Tower',
 'Tower Week',
 'Tower Tuesdays',
 'Jose

In [19]:
#Removing Duplicate Tags/Features(found due to case sensitivity) After Setting Original Tag to 1.
print('Before Duplication Hearst Magazines:', image_df_original['Hearst Magazines'].sum())
print('Before Duplication Magazines' ,image_df_original['Magazines'].sum())
print('Hearst magazines count' , image_df_original['Hearst magazines'].sum())
print('magazines count' , image_df_original['magazines'].sum())

image_df_original.loc[image_df_original['Hearst magazines']==1,'Hearst Magazines'] = 1
image_df_original.loc[image_df_original['magazines']==1,'Magazines'] = 1

print('After Duplication handling Hearst Magazines:' ,image_df_original['Hearst Magazines'].sum())
print('After Duplication handling Magazines' ,image_df_original['Magazines'].sum())

Before Duplication Hearst Magazines: 1611
Before Duplication Magazines 1494
Hearst magazines count 532
magazines count 456
After Duplication handling Hearst Magazines: 1611
After Duplication handling Magazines 1612


In [20]:
image_df_original.drop(labels=['Hearst magazines','magazines'], axis=1, inplace = True)
tag_list = [x for x in tag_list if x != 'Hearst magazines']
tag_list = [x for x in tag_list if x != 'magazines']


In [21]:

image_df_original['target'] = image_df_original.iloc[:,7:].apply(lambda row: '_'.join(row.values.astype(str)), axis=1)


In [22]:
df =pd.DataFrame(image_df_original[image_df_original['People @ Work']==1].iloc[:,8:].sum(axis = 0))
df.columns = ['counts']
df = df.reset_index()
df.head()

index counts
0            Events   5896
1        44th Floor   2989
2             Texas   2075
3  Hearst Magazines   1531
4         Magazines   1531

In [23]:
type(df['counts'].iloc[1])

numpy.int64

In [26]:
df.sort_values(by=['counts'])

TypeError: '>' not supported between instances of 'numpy.ndarray' and 'str'

In [ ]:
len(image_df_original['target'].unique())

## ETL For Final Dataset for Folder Creation

#### Write separate function or a combination of functions which follows the following steps:
1. Create folder for a tag in tag_list
2. filter images in clean_df for which contain that tag in the keyword
3. iterate through each file name in the clean_df. Concatenate with source parth and set it as src_dir
4. set destination dir as the folder created in step 1.

In [ ]:
# #Create Directory using OS

# main_dir = "C:/Examples/Python_files/OS_module"
 
# os.makedirs(main_dir,mode = 0o666) 
# print("Directory '% s' is built!" % main_dir) 

# #copy image from one folder to another in python
# src_dir = "your/source/dir"
# dst_dir = "your/destination/dir"
# for jpgfile in glob.iglob(os.path.join(src_dir, "*.jpg")):
#     shutil.copy(jpgfile, dst_dir)

##### Step 1 : Defining a function to Create a Folder for a tag list.

In [ ]:
def create_dir(image_df,tag_list):

## create folders
    for i in tag_list:
        main_dir = "C:/Users/abhij/Hearst/Train_Images/"+str(i)

        os.makedirs(main_dir,mode = 0o666) 
        print("Directory '% s' is built!" % main_dir) 

## filter dataframe on current tag to obtain image file list
        file_name_list = list(image_df_original[image_df_original[i]==1]['File Name'])
    
## iterate through the file name list to set source and destination path based on current tag and copy from source to destination
        for i in file_name_list:
            src_dir = "C:/Users/abhij/Hearst/images/"+str(i)
            shutil.copy(src_dir, main_dir)

##### Step 2 : filter images in clean_df for which contain that tag in the keyword

In [ ]:
file_name_list = list(image_df_original[image_df_original[tag_list[0]]==1]['Record ID'].str.)

In [ ]:
file_name_list

In [ ]:
# src_dir = "C:/Users/abhij/images/"+str(file_name_list[0])+".jpg"
# dest_dir = "C:/Users/abhij"
# #shutil.copy(src_dir,dest_dir)

In [ ]:
image_df_original